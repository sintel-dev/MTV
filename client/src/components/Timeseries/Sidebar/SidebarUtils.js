import * as d3 from 'd3';
import { colorSchemes } from '../FocusChart/Constants';
import { fromMonthToIndex } from '../../../model/utils/Utils';

export const getWrapperSize = () => {
  const sidebarHeight = document.querySelector('.right-sidebar').clientHeight - 40; // padding-top and bottom
  const sidebarHeaderHeight = document.querySelector('.period-control').clientHeight + 10; // margin-bottom;
  const wrapper = document.querySelector('#dataWrapper');

  const height = sidebarHeight - sidebarHeaderHeight;
  const width = wrapper.clientWidth - 7; // scrollbar offset

  return { width, height };
};

export const drawArc = (currentPeriod, periodEvents, radius, periodIndex) => {
  const { level } = currentPeriod;
  let arcStart = 0;
  let arcStop = 0;
  let base = 0;
  let arcData = [];
  const circleMonths = (2 * Math.PI) / 12;
  const circleHours = (2 * Math.PI) / 24;

  const secondsInMonth = 2629743.83;
  const secondsInDay = 86400;

  const arc = d3
    .arc()
    .innerRadius(radius - 2)
    .outerRadius(radius + 2);

  if (level === 'year') {
    const yearHasEvents = periodEvents && periodEvents[currentPeriod.name];
    if (yearHasEvents) {
      const yearEvents = periodEvents[currentPeriod.name].events;
      Object.values(yearEvents).forEach((event) => {
        const { start_time, stop_time, tag, id } = event;
        base = new Date(Number(currentPeriod.name)).getTime() / 1000;
        arcStart = ((start_time - base) / secondsInMonth) * circleMonths;
        arcStop = ((stop_time - base) / secondsInMonth) * circleMonths;
        arc.startAngle(arcStart).endAngle(arcStop);
        const tagColor = colorSchemes[tag] || colorSchemes.Untagged;
        arcData.push({ tag, tagColor, pathData: arc(), eventID: id });
      });
    }
  }

  if (level === 'month') {
    const year = currentPeriod.parent.name;
    const monthEvents = periodEvents[year] && periodEvents[year].months[fromMonthToIndex(currentPeriod.name)];
    if (monthEvents !== undefined) {
      Object.values(monthEvents.events).forEach((event) => {
        const { start_time, stop_time, tag, id } = event;
        const daysInMonth = new Date(Number(year), periodIndex + 1, 0).getDate();
        const circleDays = (2 * Math.PI) / daysInMonth;
        base = new Date(year, fromMonthToIndex(currentPeriod.name)).getTime() / 1000;
        arcStart = ((start_time - base) / secondsInDay) * circleDays;
        arcStop = ((stop_time - base) / secondsInDay) * circleDays;

        arc.startAngle(arcStart).endAngle(arcStop);
        const tagColor = colorSchemes[tag] || colorSchemes.Untagged;
        arcData.push({ tag, tagColor, pathData: arc(), eventID: id });
      });
    }
  }

  if (level === 'day') {
    const year = currentPeriod.parent.parent.name;
    const month = currentPeriod.parent.name;
    const monthNumber = fromMonthToIndex(month);
    const dayEvents =
      periodEvents[year] && periodEvents[year].months[monthNumber] && periodEvents[year].months[monthNumber].days;

    if (dayEvents !== undefined) {
      if (dayEvents[periodIndex]) {
        const { start_time, stop_time, tag, id } = dayEvents[periodIndex];
        base = new Date(year, monthNumber, periodIndex).getTime() / 1000;
        arcStart = ((start_time - base) / 3600) * circleHours;
        arcStop = ((stop_time - base) / 3600) * circleHours;
        arc.startAngle(arcStart).endAngle(arcStop);
        const tagColor = colorSchemes[tag] || colorSchemes.Untagged;
        arcData.push({ tag, tagColor, pathData: arc(), eventID: id });
      }
    }
  }
  return arcData;
};

export const getDataScale = (innerRadius, outerRadius, periodRange) => {
  const scaleAngle = d3
    .scaleLinear()
    .range([0, 2 * Math.PI])
    .domain([0, periodRange.length - 0.08]);

  const scaleRadius = d3.scaleLinear().range([innerRadius, outerRadius]).clamp(true).domain([0, 1.2]);

  const area = d3
    .areaRadial()
    .angle((d, i) => scaleAngle(i))
    .innerRadius(() => scaleRadius(0))
    .outerRadius((d) => scaleRadius(d))
    .curve(d3.curveCardinalClosed);

  const area0 = d3
    .areaRadial()
    .angle((d, i) => {
      scaleAngle(i);
    })
    .innerRadius(() => scaleRadius(0))
    .outerRadius(() => scaleRadius(0))
    .curve(d3.curveCardinalClosed);

  return { scaleAngle, scaleRadius, area, area0 };
};
