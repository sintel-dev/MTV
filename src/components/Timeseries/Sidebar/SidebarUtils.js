import * as d3 from 'd3';
import { colorSchemes } from '../FocusChart/Constants';
import { fromMonthToIndex } from '../../../model/utils/Utils';

export const MAX_EVENTS_ACTIVITY = 5;

export const getWrapperSize = () => {
  const sidebarHeight = document.querySelector('.right-sidebar').clientHeight - 75; // padding-top and bottom
  const sidebarHeaderHeight = document.querySelector('.period-control').clientHeight + 25; // margin-bottom;
  const wrapper = document.querySelector('#dataWrapper');

  const height = sidebarHeight - sidebarHeaderHeight;
  const width = wrapper.clientWidth - 22; // scrollbar offset

  return { width, height };
};

export const drawArc = (currentPeriod, periodEvents, radius, periodIndex) => {
  const { level } = currentPeriod;
  let arcStart = 0;
  let arcStop = 0;
  let base = 0;
  let arcData = [];
  const circleMonths = (2 * Math.PI) / 12;
  const circleHours = (2 * Math.PI) / 24;

  const secondsInMonth = 2629743.83;
  const secondsInDay = 86400;

  const arc = d3
    .arc()
    .innerRadius(radius - 0)
    .outerRadius(radius + 4);

  if (level === 'year') {
    const yearHasEvents = periodEvents && periodEvents[currentPeriod.name];
    if (yearHasEvents) {
      const yearEvents = periodEvents[currentPeriod.name].events;
      Object.values(yearEvents).forEach((event) => {
        const { start_time, stop_time, tag, id } = event;
        base = new Date(Number(currentPeriod.name)).getTime() / 1000;
        arcStart = ((start_time - base) / secondsInMonth) * circleMonths;
        arcStop = ((stop_time - base) / secondsInMonth) * circleMonths;
        arc.startAngle(arcStart).endAngle(arcStop);
        const tagColor = colorSchemes[tag] || colorSchemes.Untagged;
        arcData.push({ tag, tagColor, pathData: arc(), eventID: id });
      });
    }
  }

  if (level === 'month') {
    arc.outerRadius(radius + 3);
    const year = currentPeriod.parent.name;
    const monthEvents = periodEvents[year] && periodEvents[year].months[fromMonthToIndex(currentPeriod.name)];
    if (monthEvents !== undefined) {
      Object.values(monthEvents.events).forEach((event) => {
        const { start_time, stop_time, tag, id } = event;
        const daysInMonth = new Date(Number(year), periodIndex + 1, 0).getDate();
        const circleDays = (2 * Math.PI) / daysInMonth;
        base = new Date(year, fromMonthToIndex(currentPeriod.name)).getTime() / 1000;
        arcStart = ((start_time - base) / secondsInDay) * circleDays;
        arcStop = ((stop_time - base) / secondsInDay) * circleDays;

        arc.startAngle(arcStart).endAngle(arcStop);
        const tagColor = colorSchemes[tag] || colorSchemes.Untagged;
        arcData.push({ tag, tagColor, pathData: arc(), eventID: id });
      });
    }
  }

  if (level === 'day') {
    arc.outerRadius(radius + 2);
    const year = currentPeriod.parent.parent.name;
    const month = currentPeriod.parent.name;
    const monthNumber = fromMonthToIndex(month);
    const dayEvents =
      periodEvents[year] && periodEvents[year].months[monthNumber] && periodEvents[year].months[monthNumber].days;

    if (dayEvents !== undefined) {
      if (dayEvents[periodIndex]) {
        const { start_time, stop_time, tag, id } = dayEvents[periodIndex];
        base = new Date(year, monthNumber, periodIndex).getTime() / 1000;
        arcStart = ((start_time - base) / 3600) * circleHours;
        arcStop = ((stop_time - base) / 3600) * circleHours;
        arc.startAngle(arcStart).endAngle(arcStop);
        const tagColor = colorSchemes[tag] || colorSchemes.Untagged;
        arcData.push({ tag, tagColor, pathData: arc(), eventID: id });
      }
    }
  }
  return arcData;
};

export const getDataScale = (innerRadius, outerRadius, periodRange, relativaScale, currentPeriodExtent) => {
  const scaleAngle = d3
    .scaleLinear()
    .range([0, 2 * Math.PI])
    .domain([0, periodRange.length - 0.08]);

  const scaleRadius = d3.scaleLinear().range([innerRadius, outerRadius]);
  if (relativaScale) {
    scaleRadius.domain([0, currentPeriodExtent[1]]);
  } else {
    scaleRadius.domain([0, 1]);
  }

  const area = d3
    .areaRadial()
    .angle((d, i) => scaleAngle(i))
    .innerRadius(() => scaleRadius(0))
    .outerRadius((d) => scaleRadius(d))
    .curve(d3.curveCardinalClosed);

  const area0 = d3
    .areaRadial()
    .angle((d, i) => {
      scaleAngle(i);
    })
    .innerRadius(() => scaleRadius(0))
    .outerRadius(() => scaleRadius(0))
    .curve(d3.curveCardinalClosed);

  return { scaleAngle, scaleRadius, area, area0 };
};
