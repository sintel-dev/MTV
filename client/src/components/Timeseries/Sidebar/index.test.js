import React from 'react';
import { Sidebar } from './index';
import * as sidebarUtils from './SidebarUtils';
import { dataRun } from '../../../tests/testmocks/dataRun';

jest.mock('./Header', () => () => 'Header component');

describe('Sidebar Unit tests -> ', () => {
  beforeAll(() => {
    sidebarUtils.getWrapperSize = jest.fn().mockReturnValue({ witdth: 200, height: 200 });
  });
  const sidebarProps = {
    dataRun,
    experimentData: {
      isExperimentDataLoading: false,
    },
  };

  it('Should Render without crashing', () => {
    const sidebar = shallow(<Sidebar {...sidebarProps} />);
    expect(sidebar).toMatchSnapshot();
  });
});
