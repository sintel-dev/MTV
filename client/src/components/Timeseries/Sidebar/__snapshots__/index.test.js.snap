// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Unit tests ->  Should Render without crashing 1`] = `
<div
  className="right-sidebar"
>
  <Loader
    isLoading={false}
  >
    <Component />
    <div
      className="data-wrapper"
      id="dataWrapper"
    >
      <div
        className="wrapper-container scroll-style"
        style={
          Object {
            "height": "200px",
          }
        }
      >
        <svg
          height={NaN}
          id="multiPeriodChart"
        >
          <defs>
            <radialGradient
              id="blueGradient"
            >
              <stop
                offset="0"
                stopColor="#B2C1FF"
              />
              <stop
                offset="100"
                stopColor="rgba(216,216,216,0)"
              />
            </radialGradient>
          </defs>
        </svg>
      </div>
    </div>
  </Loader>
</div>
`;
